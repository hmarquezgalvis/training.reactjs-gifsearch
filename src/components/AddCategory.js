import React, { useState } from 'react'
import PropTypes from 'prop-types';

const AddCategory = ({ setCategories }) => {

  const initialState = '';
  const [ inputValue, setInputValue ] = useState(initialState)

  const handleInputChange = ( e ) => {
    setInputValue(e.target.value);
  };

  const handleSubmit  = ( e ) => {
    e.preventDefault();

    if (inputValue === '') return;

    setCategories( values => [ inputValue, ...values] );
    setInputValue( '' );
  };

  return (
    <form className="mb-3"
      onSubmit={ handleSubmit }
    >
      <p>{ inputValue }</p>
      <label className="form-label">Category:</label>
      <input className="form-control" type="text"
        value={ inputValue }
        onChange={ handleInputChange }
      />
    </form>
  )
}

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired
};

export { AddCategory }
