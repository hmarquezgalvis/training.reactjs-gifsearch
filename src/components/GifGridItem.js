import React from 'react';

import './GifGridItem.css';

export const GifGridItem = ({ id, title, url }) => {
  return (
    <>
      <div className="col-3">
        <img className="border border-white animate_animated animate__bounceIn"
          src={ url } 
          alt={ title } />
      </div>
    </>
  )
}
