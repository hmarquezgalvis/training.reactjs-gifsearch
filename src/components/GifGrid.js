import React from 'react'
import PropTypes from 'prop-types';

import { GifGridItem } from './GifGridItem';
import { useFetchGifList } from '../hooks/useFetchGifList';

const GifGrid = ({ category }) => {

  const { items: imageList, loading } = useFetchGifList( category );

  return (
    <div className="card mb-3">
      <div className="card-body">
        <h5 className="card-title">
          { category }
          { 
            loading && <div className="spinner-grow spinner-grow-sm ms-1" role="status">
                        <span className="visually-hidden">Loading...</span>
                      </div>
          }
        </h5>

        <div className="d-flex align-content-stretch flex-wrap">
          {
            imageList.map( (item, idx) =>
              <GifGridItem 
                key={ item.id }
                { ...item } />
            )
          }
        </div>
      </div>
    </div>
  )
}

GifGrid.propTypes = {
  category: PropTypes.string.isRequired
};

export default GifGrid;
