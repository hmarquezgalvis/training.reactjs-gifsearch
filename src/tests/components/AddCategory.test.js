import { shallow } from 'enzyme';
import '@testing-library/jest-dom';

import { AddCategory } from '../../components/AddCategory';

describe('tests for <AddCategory />', () => {

  const setCategories = jest.fn();
  let wrapper;

  beforeEach( () => {
    jest.clearAllMocks();
    wrapper = shallow( <AddCategory setCategories={ setCategories } /> );
  });

  test( 'must be the right html generated', () => {
    expect( wrapper ).toMatchSnapshot();
  });

  test( 'must be change the input', () => {
    const input = wrapper.find( 'input' );
    const value = 'Hi World';

    input.simulate( 'change', { target: { value } } );

    expect( wrapper.find( 'p' ).text().trim() ).toBe( value );
  });

  test( 'must be send the information on  submit', () => {
    wrapper.find( 'form' ).simulate( 'submit', { preventDefault() { } } );

    expect( setCategories ).not.toHaveBeenCalled();
  });

  test ( 'must call setCategories and clean de input text', () => {
    const value = 'Hunter x Hunter';

    wrapper.find( 'input' ).simulate( 'change', { target: { value } } );

    expect( wrapper.find( 'input' ).prop('value') ).toBe( value );

    wrapper.find( 'form' ).simulate( 'submit', { preventDefault() { } } );

    expect( setCategories ).toHaveBeenCalled();
    expect( setCategories ).toHaveBeenCalledTimes(1);
    expect( setCategories ).toHaveBeenCalledWith( expect.any(Function) );

    expect( wrapper.find( 'input' ).prop('value') ).toBe( 'no' );

  });

});