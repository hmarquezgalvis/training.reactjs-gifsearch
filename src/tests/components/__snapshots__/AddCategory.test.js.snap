// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for <AddCategory /> must be the right html generated 1`] = `
<form
  className="mb-3"
  onSubmit={[Function]}
>
  <p />
  <label
    className="form-label"
  >
    Category:
  </label>
  <input
    className="form-control"
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
